### ISSUE TITLE
Gradient Boosted Trees prediction provides wrong result if algorithm and result datatypes do not match

### ISSUE THREAD

**Issue by Mightrider** [oneapi-src/oneDAL#1002]
Hi!
This is basically the same issue [this](https://github.com/oneapi-src/oneDAL/issues/271#issue-550079036) one only with Gradient Boosted Trees.

**Describe the bug**
I want to use shallow copies of my own memory for my classification to avoid copying data. For the data input that works fine. To get the predicted results I create a prediction::ResultPtr and set a NumericalTablePtr with my own memory. If the daalDataType and the resultDataType are the same I get the expected results. If the resultDataType differs only the first element in my array is changed. The rest seems to stay the same. 

I tried the same procedure with most all the other classifiers and there the results are correct if the datatypes mismatch. 

**To Reproduce**
```c++
#include <daal.h>
#include <service.h>

int main() {
    const size_t nClasses = 2;

    typedef int dataType;
    typedef double daalDataType;

    // ## with a mismatching type (compared to daalDataType) only the first entry in vResultsMemory is changed
    typedef int resultDataType;
    // ## works as expected
    // typedef double resultDataType;


    /* creating arrays for shallow data access */
    dataType vData[8] = { 0, 0, 0, 1, 1, 0, 1, 1 };
    dataType vLabels[4] = { 0, 1, 1, 0 };

    dataType vTestData[8] = { 0, 1, 1, 0, 0, 0, 1, 1 };
    dataType vTestLabels[4] = { 1, 1, 0, 0 };

    resultDataType vResultsMemory[4] = { -2, -2, -2, -2 };

    const size_t vPrintLines = 4;

    using namespace daal;
    using namespace daal::algorithms;
    using namespace daal::data_management;
    using namespace daal::algorithms::gbt::classification;

    /* Create Numeric Tables for training data and labels with the data pointer of xtensor */
    NumericTablePtr vTrainData(HomogenNumericTable<dataType>::create(vData, 2, 4));
    NumericTablePtr vTrainGroundTruth(HomogenNumericTable<dataType>::create(vLabels, 1, 4));

    printNumericTables<dataType, dataType>(vTrainData, vTrainGroundTruth, "Train features:", "Train labels:", "Check data:", vPrintLines);

    /* Create an algorithm object to train the decision forest classification model */
    training::Batch<daalDataType> vTrainingBatch(nClasses);

    /* Pass a training data set and dependent values to the algorithm */
    vTrainingBatch.input.set(classifier::training::data, vTrainData);
    vTrainingBatch.input.set(classifier::training::labels, vTrainGroundTruth);

    /* Build the model */
    vTrainingBatch.compute();

    std::cout << "Training Complete" << std::endl;

    /**
     * Predict using the trained model
     */
    NumericTablePtr vNumTestData(HomogenNumericTable<dataType>::create(vTestData, 2, 4));
    NumericTablePtr vNumTestLabels(HomogenNumericTable<dataType>::create(vTestLabels, 1, 4));

    prediction::Batch<daalDataType> vPredictorBatch(nClasses);

    vPredictorBatch.input.set(classifier::prediction::model, vTrainingBatch.getResult()->get(classifier::training::model));
    /* Pass a testing data set and the trained model to the algorithm */
    vPredictorBatch.input.set(classifier::prediction::data, vNumTestData);

    // We want to get the result in c memory, so first we create a shallow NumericTablePtr
    NumericTablePtr vPredictedLabels(HomogenNumericTable<resultDataType>::create(vResultsMemory, 1, 4));
    // we create a new resultPtr
    classifier::prediction::ResultPtr vResultPtr(new classifier::prediction::interface2::Result());

    // and assign our shallow NumericTablePtr to the prediction
    vResultPtr->set(classifier::prediction::prediction, vPredictedLabels);
    vPredictorBatch.setResult(vResultPtr);

    /* Predict the values */
    vPredictorBatch.compute();

    /* Retrieve the algorithm results */
    printNumericTables<int, int>(vNumTestLabels, vPredictorBatch.getResult()->get(classifier::prediction::prediction),
                                 "Expected results", "Prediction results",
                                 "classification sample program results:", vPrintLines);

    // Thanks to our shallow copy we can directly access the results in vPredictedLabels
    for (size_t i = 0; i < 4; i++) {
        std::cout << vResultsMemory[i] << std::endl;
    }
}
```

**Expected behavior**
Results I get:
```
classification sample program results:
Expected resultsPrediction results
1              1              
1              -2             
0              -2             
0              -2
```

Results I expect (result for matching datatypes):
```
classification sample program results:
Expected resultsPrediction results
1              1              
1              1              
0              1              
0              1
```

**Environment:**
 - OS: [Ubuntu 18.04]
 - Compiler: [GCC7.4]
 - Version: [2020 Update 2]

**Comment by KalyanovD** [comment]
Hello @Mightrider,
Thank you for your great attention to quality of our product! You was right, this issue has the same nature as the previous one. Fix for this bug has already been prepared and merged into master branch and we are going to merge it into rls/2021-gold-mnt soon. Please, feel free to try it with upcoming 2021.1-beta10 release and contact us if you have any concerns or questions.

Thank you for valuable contribution to oneDAL project!

**Comment by emmenlau** [comment]
Thanks a lot @KalyanovD for this immediate action and positive response! Can I ask, we are still using DAAL (planning to upgrade to oneDAL at some point, but not yet). Will the fix go into any DAAL 2020.x series too?

**Comment by KalyanovD** [comment]
Hello @emmenlau,

Contribution to DAAL 2020.x line has already been finished, last update (2020 Update 3) will be available soon but this fix can't be included into this release. The closest release with this fix would be 2021.1-beta10, it is beta versions of upcoming oneDAL 2021.1 major release.
Sorry for inconvenience!

**Comment by Mightrider** [comment]
Thank your for your quick response @KalyanovD. I am happy to help.
